lib/Conversion/TritonGPUToLLVM/GatherOpToLLVM.cpp

```diff
@@ -99,7 +127,224 @@ GatherOpConversion::matchAndRewrite(GatherOp op, OpAdaptor adaptor,
   Value packed =
       packLLElements(loc, getTypeConverter(), results, rewriter, dstType);
   rewriter.replaceOp(op, packed);
-  return success();
+}
+
+// High-level description of the algorithm:
+//
+// `isWarpLocal` checks that it is possible to compute each output element
+// without data movement across warps.
+//
+// If the gather dim is `dimN`, then this means
+//
+//   ll^-1(dimN)[(block, warp)] == 0
+//
+// for both source and index tensors: moving along the gather axis does not
+// change the warp. Broadcasted layouts are not supported, so we know the
+// layouts are permutation matrices.
+//
+// We can check this with `ll((block, warp))[dimN] == 0`.
+//
+// Let `gatherCol` be a tuple of all dimensions except the gather dimension.
+// We also check that the gather columns line up the same way with respect to
+// the warp between the source and index tensors with
+//
+//   ll_src((block, warp))[gatherCol] == ll_idx((block, warp))[gatherCol]
+//
+// This means that for all index columns, the corresponding column in the source
+// tensor is owned by the same warp.
+//
+// We also check
+//
+//   ll_src(lane)[gatherCol] == ll_idx(lane)[gatherCol]
+//
+// This boils down to the fact that the algorithm essentially emits a series of
+// index shuffles for each index value owned by each thread, and then a pile of
+// selects to pick the right value. We need to figure out given an index value
+// in a particular column, what are the source register values it could read
+// from and who owns them.
+//
+// If this relationship did not hold, then the possible source registers for
+// each index value varies with the thread, meaning the value operand provided
+// to each shuffle index instruction would depend on the thread ID. This isn't a
+// big deal. It just means would have to emit a pile of selects before each
+// shuffle as well, to pick the right source register value. But we choose not
+// to handle this.
+//
+// The codegen algorithm emits code:
```

Comment:
I didn't go through the detailed algorithm. I think the description is clear to me but I have concerns about the performance.

The idea is that we use the value in the `index` tensor to get where the values should be gathered from. Since we know that values must come from the same warp, we get the corresponding `src_reg` and `src_tid` using the inversion of the source tensor (I do think we can handle broadcast cases). Though `src_tid` is determined, we cannot control what register is sent from `src_tid`. So the brute force way here tries to get all the available `src_reg`s from the `src_tid`. **I'm not sure if it's always more efficient than the shared memory-based approach.**

I had a discussion with Justin long time ago, if we could construct a layout that has `dst_reg` and `src_tid`, then the algorithm complexity could be reduced. (I didn't dive into it though)

Reply:
The problem comes from the data dependency of the shuffle. This algorithm makes the thread owning the index handle all data dependency, because other threads could want to read different registers from the same source thread.

There is an alternative implementation that scales better using butterfly shuffles, but I haven't implemented it in this PR. The idea is emitWarpLocalGather would compute how many shuffles each will emit, and at a certain numRegsPerColumn, the butterfly shuffle version is better.

In any case, the op will also rely on the middle end picking the best layout

Reply:
I think it might be good to benchmark the perf a bit. I'm not saying that a bit slowness will prevent us from integrating this implementation, but if we see very significant drop comparing to shared memory version (though not likely), may need to investigate.

Reply:
I think it's a bit premature to concentrate too much on performance numbers here. We know that in some cases the shuffle path will be the most efficient so we we want to have this path working. 
We will need more real life data to decide on what cases are interesting and how to pick the heuristic for those. As Jeff mentioned this will also be highly depend on how we pick the layout and propagate it.

I don't think the butterfly shfl would make a big difference here but I could be wrong.

Reply:
IMO it's fine to merge since Jeff's perf numbers seem reasonable anyway

Reply:
I documented the current limitations of this codegen path.